Add content type route matcher

ParisLiakos opened the pull request:
Useful for REST APIs that accept a POST or PUT request with content type requirements.

bradleybeddoes commented on gotham/src/router/route/matcher/content_type.rs:
```diff
@@ -0,0 +1,92 @@
+    /// Determines if the `Request` was made using a `Content-Type` header that includes one or more
+    /// supported media types. A missing `Content-Type` header, or the value of `*/*` will not match.
+    fn is_match(&self, state: &State) -> Result<(), RouteNonMatch> {
+        // Request method is valid, ensure valid Accept header
```
Just a minor invalid comment (which I note needs fixing in AcceptHeaderRouteMatcher too)

bradleybeddoes commented on gotham/src/router/route/matcher/content_type.rs:
```diff
@@ -0,0 +1,92 @@
+        let headers = Headers::borrow_from(state);
+        match headers.get::<ContentType>() {
+            Some(content_type) => {
+                if self.supported_media_types.contains(&content_type.0) {
```
Forgive me, I have a head cold so my brain is not working well, but I'm just not seeing how the `.0` accessor is being used successfully here? (obviously it is!)

ParisLiakos replied:
If you see Content type's signature, it is just wrapping a Mime
`pub struct ContentType(pub Mime);`
Its `Display` implementation also uses the same way to access the Mime, see
https://docs.rs/hyper/0.11.24/src/hyper/header/common/mod.rs.html#306

bradleybeddoes commented on gotham/src/router/route/matcher/content_type.rs:
```diff
@@ -0,0 +1,92 @@
+use state::{request_id, FromState, State};
+
+/// A `RouteMatcher` that succeeds when the `Request` has been made with a `Content-Type` header
+/// that includes one or more supported media types. If the Content-Type is missing
```
worth noting that `*/*` fails as well (as you have later on)

bradleybeddoes commented on gotham/src/router/route/matcher/content_type.rs:
```diff
@@ -0,0 +1,92 @@
+/// // No content type header
+/// state.put(Headers::new());
+/// assert!(matcher.is_match(&state).is_err());
+///
```
Suggest adding example showing `*/*` being in error also

ParisLiakos commented:
Thanks for the review! i implemented your suggestion but then realized that wildcard `Content-Type` makes no sense. I googled around and also found this; 
https://www.w3.org/Protocols/HTTP/HTRQ_Headers.html

Probably having wildcards in it is invalid.
So, i removed references to wildcards and further adjusted docs

bradleybeddoes commented:
Failures due to travis timeouts, not issues with this code per se, approving.
